File: server/controllers/workers/barCodes.ts
@Namespace('bar-codes')
export class BarCodesController {
  @Model('BarCodesModel') private barCodesModel: Models<'BarCodesModel'>

  @On('create')
  public async create(newBarCode: Miscellaneous.NewBarCode): Promise<any> {
    const { name, tag = '', value } = newBarCode
    await this.barCodesModel.create({ name, tag, value })
    return {
      ok: true
    }
  }

  @On('get')
  public get(): Promise<any> {
    return this.barCodesModel.getAll()
  }

  @On('update')
  public async update(barCode: Miscellaneous.BarCode): Promise<any> {
    const { id, name, tag = '', value } = barCode
    await this.barCodesModel.update(id, { name, tag, value })
    return {
      ok: true
    }
  }

  @On('delete')
  public async delete(id: Miscellaneous.BarCode['id']): Promise<any> {
    await this.barCodesModel.delete(id)
    return {
      ok: true
    }
  }
}

File: core/models/barcodes.go
package models

type BarCode struct {
	Id    string `json:"id"`
	Name  string `json:"name"`
	Tag   string `json:"tag"`
	Value string `json:"value"`
}

type NewBarCode struct {
	Name  string `json:"name"`
	Tag   string `json:"tag"`
	Value string `json:"value"`
}

type BarCodesModel interface {
	Create(data NewBarCode)
	Update(data BarCode)
	Get(id string) *BarCode
	GetAll() []BarCode
	Delete(id string)
}
